qt: fix broken unicode chars on osx 10.10

theuni opened the pull request:
Fixes #5657 and likely #5165.

The default font changed again.

The real fix is to compile qt against a >= 10.8 sdk, but this is simple enough
to backport to 0.10 to avoid having to do that there.

Note: NSAppKitVersionNumber is a double and there's no official value for
NSAppKitVersionNumber10_10. Since == isn't reliable for doubles, use Apple's
guidelines for testing versions here:
https://developer.apple.com/library/mac/releasenotes/AppKit/RN-AppKit/

theuni committed:
qt: fix broken unicode chars on osx 10.10

The default font changed again.

The real fix is to compile qt against a >= 10.8 sdk, but this is simple enough
to backport to 0.10 to avoid having to do that there.

Note: NSAppKitVersionNumber is a double and there's no official value for
NSAppKitVersionNumber10_10. Since == isn't reliable for doubles, use Apple's
guidelines for testing versions here:
https://developer.apple.com/library/mac/releasenotes/AppKit/RN-AppKit/

Chinese and Japanese fonts have been hard-coded as well, otherwise they fail to
show up at all.

theuni commented:
Unfortunately this doesn't fix #5165. I spent most of the day chasing my tail on this and got nowhere. Unless someone else is interested in looking into this, we have 2 options for 0.10: busted Chinese for Yosemite, or building against the 10.8 sdk.

jonasschnelli commented:
Will test this. Can I just gbuid master?

Busted Chinese on Yosemite 10.10 sounds more sane than switching to 10.8 in rc4 IMO.

theuni commented:
Uploading a binary, 2min.

theuni commented:
@jonasschnelli https://bitcoincore.org/cfields/bitcoin-qt-10.6_quickfix2-9ca6e0479 . I'm inclined to agree.

jonasschnelli commented:
Tested. Fixes #5657 but leaves #5165 open.

The problem of #5165 is, that the 'Lucia Grande' font does not support Chinese characters. We now substituting the system font by a non-Chinese-capable  font.

I think a quick fix could be to check within the 10.10 if the user-language is zh_CN (maybe there are others?) then substitute the font to "Henti TC" (OS X standard font for Chinese). A one or two-liner with low risks (only Chinese 10.10 users will see the change).
Easy, sane, effective IMO.

theuni commented:
@jonasschnelli I wasn't sure if we could do that, but since we require a restart to change fonts, that sounds good to me.

It seems that all cjk fonts are broken in 10.10 due to the new aliasing system. Hard-coding for those 3 indeed sounds like a reasonable hack for 0.10.

laanwj commented:
If #5165 only affects Chinese, then this + a Chinese-specific workaround for 0.10 is great. 
I hope other languages with non-latin characters aren't affected, tracking down the font for each locale would be a crazy scavenger hunt.

And indeed, for master we shouldnt bother with this and switch the build SDK.

jonasschnelli commented:
The current substitute font 'Lucida Grand' on osx 10.10 has only support for Greek, Hebrew, Cyrillic, Latin.

So in general we should consider not giving substitutes for fonts because i assume the OS will care about switching fonts within a textflow or support extra fonts with mixed charsets.

Currently the 'Lucida Grand' font has no support for:
Hiragana, Birman, Arabic, Hangul, Thai, Traditional Chinese, Simple Chinese, Singhalen, Yi (maybe more)
Do we support a language with one of these font script?

ACK on this change to fix #5165 but longterm we should update to 10.8 as build os and remove font substitution.

![bildschirmfoto 2015-01-16 um 10 47 39](https://cloud.githubusercontent.com/assets/178464/5774425/ea9ed802-9d6d-11e4-9315-1f33d384ec1c.png)

theuni committed:
qt: fonts: allow SubstituteFonts to filter based on user's language

SubstituteFonts() has been moved to after app identification so that QSettings
are accessible.

theuni committed:
qt: avoid hard-coding font names

They may not contain all necessary characters for a language

theuni commented:
Well that took quite a bit of trial+error, but all good now. Fixes #5165. It's not optimal... Chinese font-names don't show in English (when you're in English and going to select Chinese/Japanese from the menu, those characters are blocks). After switching it works fine though.

The last commit avoids the hard-coded use of Arial, which does not always work correctly with Chinese. Seems sane to me, but I'm not sure if there was a specific reason for its use.

The substitutions are all wrapped up in ifdefs, so they're safe for master now and in the future. The only functional change for master (after we bump the sdk to 10.8) is the Arial change.

Test binary here: https://bitcoincore.org/cfields/bitcoin-qt-10.6_quickfix3-437de97c
As far as I'm concerned, ready for merge/backport.

theuni commented:
If it wasn't clear, the Arial change wasn't arbitrary... the splash is all boxes in Chinese otherwise.

jonasschnelli commented on src/qt/guiutil.cpp:
```diff
@@ -393,12 +396,28 @@ void SubstituteFonts()
+            else if (language == "ja") // Japanesee
+              QFont::insertSubstitution(".Helvetica Neue DeskInterface", "Songti SC");
+            else
+              QFont::insertSubstitution(".Helvetica Neue DeskInterface", "Helvetica");
```
Why we substitute the default Font with `Helvetica`?
Looks good but leads to new problems (see other comment).

![bildschirmfoto-2015-01-17-um-08 45 03](https://cloud.githubusercontent.com/assets/178464/5788120/76dbadc4-9e25-11e4-84f5-bf9e19f45251.png)

jonasschnelli commented:
Still issues (see screens below).
I would even say it's more broken then before this pull (with all respect of your highly a valued work).

I only see one direction:
-> build against 10.8 with a -min of 10.6.
Im pretty sure QT won't use any methods not available in 10.6. The native osx implementation within the #ifdefs are sane because i did the most and i did always check during runtime if a method is available before calling it.
Of course it's a hard admission in rc3 to change the building framework. But IMO it should be sane.
I could also offer extensive testing in 10.6, 10.7, 10.8, 10.9 and 10.10.

zh_CH:
![bildschirmfoto-2015-01-17-um-08 40 05](https://cloud.githubusercontent.com/assets/178464/5788122/8bb3b87c-9e25-11e4-9a4e-31aba25ca402.png)

de_DE:
![bildschirmfoto-2015-01-17-um-08 42 29](https://cloud.githubusercontent.com/assets/178464/5788124/8ee41eba-9e25-11e4-94d4-ffffed27e1d1.png)

ja:
![bildschirmfoto 2015-01-17 um 08 50 05](https://cloud.githubusercontent.com/assets/178464/5788135/ef761b7a-9e25-11e4-9f4e-a971443002e9.png)


Font mix in de_DE:
![bildschirmfoto 2015-01-17 um 08 40 41](https://cloud.githubusercontent.com/assets/178464/5788133/e8107290-9e25-11e4-90fd-0746f363e1f1.png)

jonasschnelli commented:
Tested ACK.
Still have minor issues in CN/JA. This will be fixed in the next release when we might upgrade to 10.9 as build os.

Screens:
![bildschirmfoto 2015-01-17 um 09 36 19](https://cloud.githubusercontent.com/assets/178464/5788291/9d796a78-9e2c-11e4-8501-599aeefc66dd.png)

![bildschirmfoto 2015-01-17 um 09 36 31](https://cloud.githubusercontent.com/assets/178464/5788292/a0f7f7dc-9e2c-11e4-8069-e88a1e8a7191.png)

![bildschirmfoto 2015-01-17 um 09 36 34](https://cloud.githubusercontent.com/assets/178464/5788293/a47c142e-9e2c-11e4-8389-f43493b4db77.png)

![bildschirmfoto 2015-01-17 um 09 36 42](https://cloud.githubusercontent.com/assets/178464/5788294/abfc06fa-9e2c-11e4-8d7a-697799784fb1.png)

![bildschirmfoto 2015-01-17 um 09 36 53](https://cloud.githubusercontent.com/assets/178464/5788295/b0621b62-9e2c-11e4-8843-4d99316e0f2f.png)

![bildschirmfoto 2015-01-17 um 09 36 56](https://cloud.githubusercontent.com/assets/178464/5788296/b4104fa4-9e2c-11e4-86ba-8a9a5bd3891f.png)

![bildschirmfoto 2015-01-17 um 09 37 06](https://cloud.githubusercontent.com/assets/178464/5788297/b90407bc-9e2c-11e4-842b-850e7ba387ff.png)

![bildschirmfoto 2015-01-17 um 09 37 08](https://cloud.githubusercontent.com/assets/178464/5788298/bd7ed02e-9e2c-11e4-9e0e-2a7ac04957a8.png)

theuni commented:
After some discussion on IRC, I believe @jonasschnelli and I have reached a compromise here. German is fixed up now, and Chinese and Japanese work with the exception of a few artifacts. We're willing to accept that for 0.10, with the understanding that they'll be fixed in the future by the bumped sdk.

laanwj commented:
Cherry-picked to 0.10 branch as 58259ad1edf14a4f514fe5f837ac7d1d99b58422
